### PR TITLE
Memoize library

### PR DESCRIPTION
**Pull request by EvandroLG**
An implementation of the memoize technique using `Object`.

### CODE REVIEW

File: packages/function-memoize/index.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+function memoize(callback, resolver) {
+  if (resolver !== undefined && typeof resolver !== 'function') {
+    throw new Error('`resolver` should be a function');
+  }
+
+  var cache = {};
+
+  var memoized = function() {
+    var args = arguments;
+    var key = resolver ? resolver.apply(this, args) : args[0];
```
Comment (angus-c) [inline packages/function-memoize/index.js:35]: Can we improve on first argument as default key? How about JSON.stringifying argument array for key if no resolver supplied?

File: test/function-memoize/index.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+  var result = calc(2);
+
+  t.deepEqual(calc.cache, {'2': 3});
+  t.equal(result, 3);
+
+  t.end();
+});
+
+test('memoize with two parameter', function(t) {
```
Comment (angus-c) [inline test/function-memoize/index.js:21]: "parameters"

File: test/function-memoize/index.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  t.deepEqual(calc.cache, {});
+  var result = calc(5, 5);
+
+  t.deepEqual(calc.cache, {'5-5': 10});
+  t.equal(result, 10);
+
+  t.end();
+});
+
+test('get result from cache using customized key', function(t) {
```
Comment (angus-c) [inline test/function-memoize/index.js:39]: "customized key" -> custom resolver

File: README.md [hunk 4]
```diff
@@ -1143,6 +1144,37 @@ median(["3", 2, 1]); // throws
...
+
+const sumByOne = memoize(function(value) {
+  return value + 1;
+});
+
+sumByOne(10);
+sumByOne(10); -- Cache hit!
+
+sumByOne(20);
+sumByOne(20); -- Cache hit!
```
Comment (angus-c) [inline README.md:1164]: Also show it throws if arguments are not expected types

File: packages/function-memoize/index.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  })
+
+  sum(10, 10)
+  sum(10, 20)
+  sum(10, 20) -- Cache hit!
+*/
+
+function memoize(callback, resolver) {
+  if (resolver !== undefined && typeof resolver !== 'function') {
+    throw new Error('`resolver` should be a function');
```
Comment (angus-c) [inline packages/function-memoize/index.js:28]: Can we check callback is a function too, and throw if not?

**Review by angus-c — COMMENTED** [review]
Many thanks for this! Will be a good addition

File: packages/function-memoize/index.js (hunk 1 above)
Comment (EvandroLG) [reply packages/function-memoize/index.js]: Liked the idea of using `JSON. stringifying`.

File: test/function-memoize/index.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+  t.deepEqual(concat.cache, {'myKey': 'abc'});
+  t.equal(result, 'abc');
+
+  var result2 = concat('d', 'e', 'f');
+  t.deepEqual(concat.cache, {'myKey': 'abc'});
+  t.equal(result2, 'abc');
+
+  t.end();
+});
+
```
[the remark this answers is not in the document]
Comment (EvandroLG) [reply test/function-memoize/index.js:60]: This test was already included, it's the first one.

**Pull request merged**